### PR TITLE
fix(examples/vit): Make the loss actually decrease

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Fix a cluster of issues that prevented `ezpz.examples.vit` from training: AdamW was using its default `lr=1e-3` (way too high for from-scratch ViTs), no LR schedule, no proper weight init, no `qkv_bias`, and a 24-layer ViT-Large defaulting to MNIST.

### What was broken

- **`vit.py:580`**: `AdamW(model.parameters())` — no LR, no weight decay
- **`_init_weights`**: only initialized `pos_embed` and `cls_token`; every Linear used PyTorch's `kaiming_uniform` default (calibrated for ReLU-MLPs, suboptimal for ViTs)
- **`AttentionBlock.qkv`**: `bias=False` (standard ViT uses `bias=True`)
- **`--warmup`**: only gated metric collection; the optimizer slammed in at full LR from step 0
- **MNIST defaults**: depth=24, heads=16, head_dim=64 (~73M params) for 10-class MNIST, with `--max_iters=100` (~0.2 epochs)

### Fixes (2 atomic commits)

1. **`c4cfd54`** — architecture correctness: trunc_normal/xavier weight init for Linear/Conv2d, `qkv_bias=True`
2. **`85570ff`** — trainability: `--lr` (3e-4), `--weight-decay` (0.05), `--lr-warmup-iters` (5%), `--min-lr-ratio` (0.1) flags wired to a real `LambdaLR` (linear warmup → cosine decay); MNIST now picks a small ViT (depth=4, ~770K params) with `--max_iters=2000`

## Verification

500-iter CPU smoke run on MNIST:

| iter | loss | acc | lr |
|---|---|---|---|
| 0 | 2.3015 | 14% | 2e-05 (warming up) |
| 50 | 1.9770 | 25% | 5e-04 (peak) |
| 250 | 1.0883 | 66% | 3.1e-04 (cosine) |
| 500 | 0.5212 | 78% | 5e-05 (final) |

Loss crosses `log(10)≈2.30` within ~10 steps and reaches 0.5 by 500.

## Test plan

- [x] 638/638 existing tests pass (no vit-specific tests existed)
- [x] CPU smoke run on MNIST — loss drops from 2.30 → 0.52, acc 14% → 78%
- [ ] GPU/XPU run on a real allocation to verify the schedule + bf16 path on accelerators

## Summary by Sourcery

Improve the ViT example’s default configuration so the model reliably trains on MNIST, aligning it with standard ViT training practices.

New Features:
- Add configurable learning rate, weight decay, LR warmup duration, and minimum LR ratio flags wired into a linear-warmup–cosine-decay schedule for the ViT example.

Enhancements:
- Adjust MNIST ViT defaults to a smaller architecture and longer training run so default executions meaningfully reduce loss.
- Adopt a ViT-appropriate weight initialization scheme for embeddings, patch projection, and transformer Linear/LayerNorm layers.
- Enable configurable QKV bias in attention blocks and default it to match standard ViT behavior.

### CODE REVIEW

**Comment by saforem2** [comment]
Addressed all review feedback in commits 711d7a8 and 3de8cc2:

**From sourcery-ai:**
- ✅ `qkv_bias` moved to end of signature + keyword-only via `*` (711d7a8)
- ✅ `lr_warmup_iters` clamped to `max_iters - 1` so cosine decay always has at least one step (3de8cc2)
- ✅ Extracted `_compute_lr_warmup_iters` and `build_lr_scheduler` to module level for testability (3de8cc2)

**From chatgpt-codex-connector:**
- ✅ MNIST defaults no longer clobber `--model` preset fields — `apply_dataset_overrides` now skips fields the preset already set (3de8cc2)
- ✅ LR metric off-by-one fixed: capture `current_lr` before `lr_scheduler.step()` so the logged LR matches the loss it produced (3de8cc2)
  - Note: re-tested PyTorch's `LambdaLR(last_epoch=-1)` — it *does* apply `lambda(0)` to the optimizer before any `scheduler.step()` runs, so the first `optimizer.step()` is at the warmup LR, not at full `args.lr`. The actual bug was the metric-vs-actual mismatch, which is now fixed.

**From copilot-pull-request-reviewer:**
- ✅ Same first-step concern as Codex above — the actual update path was correct, the metric was off
- ✅ `--lr-warmup-iters=0` now correctly disables warmup instead of being coerced to 1 (3de8cc2)

**Tests added** (`tests/test_vit_helpers.py`, 22 new):
- `_compute_lr_warmup_iters`: zero, fraction, absolute, clamping
- `build_lr_scheduler`: warmup-reaches-peak, disabled-warmup-starts-at-peak, decays-to-min-ratio, warmup-longer-than-run-clamped, min-lr-ratio-clamping
- `apply_dataset_overrides`: MNIST default applies, model preset wins over MNIST default (regression test for the Codex finding), explicit flag wins, non-MNIST no-op
- `AttentionBlock` signature: format-still-third-positional (regression test for the qkv_bias finding), qkv_bias-keyword-only-rejects-positional, qkv_bias=False

Full suite: 638 → 660 tests passing.

File: src/ezpz/examples/vit.py [hunk 1]
```diff
@@ -577,7 +653,53 @@ def train_fn(
...
+    # Use the standard ViT recipe explicitly: AdamW(lr, weight_decay,
+    # betas=(0.9, 0.999)) — defaults gave us lr=1e-3 which routinely
+    # diverges or stalls a from-scratch ViT.
+    optimizer = torch.optim.AdamW(  # type:ignore[arg-type]
+        model.parameters(),
+        lr=args.lr,
+        weight_decay=args.weight_decay,
+        betas=(0.9, 0.999),
+    )
+    # Linear warmup → cosine decay to args.lr * args.min_lr_ratio over
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/vit.py:741]: Done in 3de8cc2 — extracted `_compute_lr_warmup_iters` and `build_lr_scheduler` to module level.  `train_fn` now just wires them up.  Also added 22 unit tests in `tests/test_vit_helpers.py` that exercise the helpers in isolation (warmup-as-fraction-vs-steps, warmup > max_iters clamping, disabled warmup, cosine floor).

File: src/ezpz/examples/vit.py [hunk 2]
```diff
@@ -128,14 +128,29 @@
...
     "img_size": 28,
     "num_classes": 10,
     "patch_size": 4,
+    # MNIST is tiny.  The ViT-Large-ish defaults (depth=24, heads=16,
+    # head_dim=64 → embed_dim=1024, ~73M params) overfit before they
+    # learn anything useful.  Use a smaller, faster model by default
+    # so a single run actually trains.
+    "num_heads": 4,
+    "head_dim": 32,
+    "depth": 4,
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/vit.py:141]: Fixed in 3de8cc2 — `apply_dataset_overrides` now skips any field that the chosen `--model` preset already set.  Added a regression test `test_model_preset_wins_over_mnist_default` in `tests/test_vit_helpers.py` that fails on the old behavior.

File: src/ezpz/examples/vit.py [hunk 3]
```diff
@@ -640,6 +762,8 @@ def train_fn(
         ezpz.distributed.synchronize()
         t3 = time.perf_counter()
         optimizer.step()
+        if lr_scheduler is not None:
+            lr_scheduler.step()
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/vit.py:837]: Partially — re-tested PyTorch's `LambdaLR(last_epoch=-1)` and it actually applies `lambda(0)` to `optimizer.param_groups` *before* training starts, so the first `optimizer.step()` IS at the warmup LR (not at `args.lr`).  Verified with a 4-line repro:

```python
opt = AdamW(m.parameters(), lr=1.0)
sch = LambdaLR(opt, lr_lambda=lambda s: 0.1*(s+1))
print(opt.param_groups[0]['lr'])  # → 0.1, not 1.0
```

The actual bug your comment helped surface was the **`train/lr` metric being off-by-one** — we logged `optimizer.param_groups[0]["lr"]` after `lr_scheduler.step()`, so each row showed the next step's LR alongside this step's loss.  Fixed in 3de8cc2 by capturing `current_lr` *before* the scheduler step.  New regression test `test_warmup_reaches_peak` checks the actual LR sequence.

[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/vit.py:837]: Same as Codex's comment above — the actual update path was correct (`LambdaLR` initializes `opt.param_groups[0]['lr']` to `peak * lambda(0)` before training), but the `train/lr` metric was logging the next step's LR.  Fixed in 3de8cc2 by capturing `current_lr` before `lr_scheduler.step()`.

Verified post-fix on a 200-iter smoke run:
```
iter=  0  loss=2.32  lr=2.5e-05  ← warmup start (1/20 × 5e-4)
iter= 20  loss=2.11  lr=5.0e-04  ← peak, exactly when warmup completes
iter=200  loss=1.35  lr=5.0e-05  ← cosine decay floor
```

File: src/ezpz/examples/vit.py [hunk 4]
```diff
@@ -577,7 +653,53 @@ def train_fn(
...
+    if args.max_iters is not None and args.max_iters > 0:
+        total_iters = int(args.max_iters)
+        if 0.0 < args.lr_warmup_iters < 1.0:
+            lr_warmup_iters = max(1, int(args.lr_warmup_iters * total_iters))
+        else:
+            lr_warmup_iters = max(1, int(args.lr_warmup_iters))
+        min_lr_ratio = max(0.0, min(1.0, args.min_lr_ratio))
+
+        def _lr_lambda(step: int) -> float:
+            if step < lr_warmup_iters:
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/vit.py]: Fixed in 3de8cc2 — `_compute_lr_warmup_iters` now handles `<= 0` explicitly and returns 0, and `build_lr_scheduler` checks `warmup_iters > 0` before applying the linear-warmup branch.  New test `test_zero_disables_warmup` and `test_disabled_warmup_starts_at_peak` cover both paths.
